### PR TITLE
Simplify quote values

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

See #5252.

Currently there are `isEqual` checks on a React objects every time a quote block re-renders. This branch avoids these isEqual checks in the `RichText` component by changing the quote values to not get recalculated every time `render` is called.

## How Has This Been Tested?
* Make sure quotes render and transformations work.
* Make sure you can move the caret to both fields in the quote.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style.
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Comment by iseulde** [comment]
@youknowriad @aduth Not sure if the structure of the value attributes had a reason or if it was just so to work with `hpq`.

**Comment by iseulde** [comment]
Also makes me wonder if we should/could warn block authors if they do something similar (return new arrays/objects on every render).

File: blocks/library/pullquote/index.js [hunk 1]
```diff
@@ -118,9 +104,7 @@ export const settings = {
 
 		return (
 			<blockquote className={ `align${ align }` }>
-				{ value && value.map( ( paragraph, i ) =>
-					<p key={ i }>{ paragraph.children && paragraph.children.props.children }</p>
-				) }
+				{ [ value ] /* Prevent keys warning... */ }
```
Comment (iseulde) [inline blocks/library/pullquote/index.js]: Not sure if we should do something else here... :)

**Comment by youknowriad** [comment]
Nice exploration :), I like these improvements.

> Not sure if the structure of the value attributes had a reason or if it was just so to work with hpq.

The structure was not necessary, it was there because the idea was that you can use any submatcher to your `query` matchers. In this PR, you forced the `node` sub matcher as default submatcher for `query`, I'm fine with that personally.

**Comment by aduth** [comment]
>In this PR, you forced the `node` sub matcher as default submatcher for `query`, I'm fine with that personally.

I'm not as fine with the change. Notably the fact that it breaks the gallery block which leverages the generalized behavior of `query`.

**Comment by aduth** [comment]
Over the weekend I started on a branch to explore eliminating `children` and `node`, which may offer similar simplifications and otherwise conflict with this proposal.

**Comment by iseulde** [comment]
@aduth So what would you propose instead? I don't really care that much _how_ the array of elements is created, just that it's not this weird structure with `[ { children: {} } ]`.

**Comment by youknowriad** [comment]
> I'm not as fine with the change. Notably the fact that it breaks the gallery block which leverages the generalized behavior of query.

Not sure I understand, the generalized behavior is still here, this PR introduces a new behavior in the `query` source if you omit the `query` attribute. Another option would be a `matcher: { type: 'node' }` to make it generic.

If you have `query` you create an object if you don't have `query` you use the `matcher`. I believe we already had this discussion before @aduth  :) (I wanted to introduce an `object` source at that time :)

**Comment by youknowriad** [comment]
related https://github.com/WordPress/gutenberg/pull/2854

**Comment by aduth** [comment]
Actually, now that we're talking about it, I think the best simplification would be to upgrade the block's to use block nesting.

**Comment by iseulde** [comment]
Okay, how do me move this forward? Make it nested? Is this a valuable intermediate step or shall I close this PR. What I'd like is to get rid of the isEqual check though which is there only for the quote black because it's calculating new values all the time.

**Comment by iseulde** [comment]
I added a message for block authors so they don't do the same... E.g. replace a RichText value prop with `[ ...value ]` and you'll see the warning:

screen shot 2018-03-06 at 23 44 52

**Comment by aduth** [comment]
I'm not a fan of the relationship between `query` and `node` that's being introduced here. My opinion is to abandon it and move forward with nested work.

**Comment by iseulde** [comment]
Continuing in #5265.

**Pull request closed without merging**